### PR TITLE
[com_fields] Store fields in 'com_fields' property instead of reusing 'params'

### PR DESCRIPTION
**Pull request by Bakual**
This decouples a part of https://github.com/joomla/joomla-cms/pull/13182 since we don't seem to get an agreement there anytime soon 😄 

### Summary of Changes
* Store fields in 'com_fields' property instead of reusing the already existing 'params'. This eliminates possible conflicts.
* Show custom fields in frontend article edit using the joomla.edit.params JLayout.
* Adjusted the JLayout to take a custom tab name so it is reusable better. If no tab name is specified it uses the current behavior ('myTab').

### Testing Instructions
* Test that you can still save values into custom fields both in frontend and backend.
* Inspect the HTML code and see that the ID attribute of the field has changed from `jform_params_alias`  to `jform_com_fields_alias` and the name from `jform[params][alias]` to `jform[com_fields][alias]`.

### Documentation Changes Required
None

### CODE REVIEW

**Comment by ralain** [comment]
I have tested this item :white_check_mark: successfully on c8ff6d1716fc17435a5938bc32a1bea9dfa566ad<br /><br />(Assuming you mean `jform[com_fields][alias]`)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13353">issues.joomla.org/tracker/joomla-cms/13353</a>.</sub>

**Comment by Bakual** [comment]
> (Assuming you mean jform[com_fields][alias])

Of course! Adjusted the PR description now 😄

**Comment by laoneo** [comment]
I don't se the benefit for this change honestly? Params is made for custom fields, so why not reuse it?

**Comment by Bakual** [comment]
> Params is made for custom fields, so why not reuse it?

Because it's not made for custom fields. It's a groupd specified in many (if not most) extensions XML file to hold various item options.
That means in most cases there is already data in that group, and it's the reason you load the existing data, convert it, merge with com_fields data, process it, unset it and convert the original params data back. All while having the risk that we have conflicts with existing data.
Now if using our own group, we can eliminate the risk of having conflicts and don't need to convert and merge and convert back the existing data.

File: plugins/system/fields/fields.php [hunk 1]
```diff
@@ -58,43 +58,24 @@ public function onContentBeforeSave($context, $item, $isNew)
 			return true;
 		}
 
-		$params = new Registry;
-
-		// Load the item params from the request
+		// Load the fields data from the request
 		$data = JFactory::getApplication()->input->post->get('jform', array(), 'array');
```
Comment (laoneo) [inline plugins/system/fields/fields.php]: Fetching the data from the post request was only done for some exception I can't remember actually. IMO the custom fields values should be fetched from the item which get saved and not from the post request. That looks for me wrong.

Comment (Bakual) [inline plugins/system/fields/fields.php]: I haven't changed that specific line, only the comment tied to it.
May it be that the custom fields data has been sanitised and removed prior to that step and thus we need to fetch it again? I haven't looked closer at it when I did this PR.

Comment (laoneo) [inline plugins/system/fields/fields.php]: In most cases, the fields values are in $data['params'] which are validated by the form. The code is on line 68.

**Comment by ggppdk** [comment]
Besides the obvious reason to avoid conflicts, there are more benefits

- since anyway they need to be seperated from params array, why add them there in the first place,

thus avoid the extra step to find the field them inside the 'params' and then clean-up the 'params' array before DB save

- it makes easier to reference / handle them inside form via JS

- easier to apply "different" validation of them ... in future if needed

- also logically it makes sense since the 'params' are mostly non-content, and 'com_fields' are mostly 'content' ...

File: plugins/system/fields/fields.php (hunk 1 above)
Comment (Bakual) [inline plugins/system/fields/fields.php]: I had a look now, the line is needed.
The thing is that `$item` is a JTable object and thus only contains the data that is going to be stored into the respective database table. There is neither a `params` or `com_fields` property at this point. So we need to fetch the raw form data and take the custom fields out of it.

Comment (laoneo) [inline plugins/system/fields/fields.php]: For contacts the $item->params property exists which holds the custom fields values. But you are right for articles not. What is missing here is the form validation when take it directly from the input.

Comment (Bakual) [inline plugins/system/fields/fields.php]: True that. I'd say it's out of scope of this PR as this behavior exists with or without this PR.

**Comment by laoneo** [comment]
We have conflicts here.

**Comment by Bakual** [comment]
Rebased and conflict solved.

**Comment by Bakual** [comment]
Rebased and fixed conflicts (again)

**Comment by Bakual** [comment]
Rebased and fixed conflicts once more

File: plugins/system/fields/fields.php [hunk 2]
```diff
@@ -31,61 +31,6 @@ class PlgSystemFields extends JPlugin
...
-	 * @param   stdClass  $item     The item
-	 * @param   boolean   $isNew    Is new
-	 *
-	 * @return  boolean
-	 *
-	 * @since   3.7.0
-	 */
-	public function onContentBeforeSave($context, $item, $isNew)
-	{
-		if (!isset($item->params))
```
Comment (laoneo) [inline plugins/system/fields/fields.php]: Please don't delete that code. It cleans up the items which do have a params attribute. Otherwise the value of the fields are saved in the params array of the item as well.

Comment (Bakual) [inline plugins/system/fields/fields.php]: The whole point of this PR is to **not** store the field data in the params property. So no, they aren't saved there we don't populate it into params anymore.  And thus that code isn't needed anymore.

Comment (laoneo) [inline plugins/system/fields/fields.php]: Yea true.

File: components/com_content/views/form/tmpl/edit.php [hunk 3]
```diff
@@ -14,6 +14,8 @@
 JHtml::_('behavior.formvalidator');
 JHtml::_('formbehavior.chosen', '#jform_catid', null, array('disable_search_threshold' => 0));
 JHtml::_('formbehavior.chosen', 'select');
+$this->tab_name = 'com-content-form';
+$this->ignore_fieldsets = array('image-intro', 'image-full', 'jmetadata', 'item_associations');
```
Comment (yvesh) [inline components/com_content/views/form/tmpl/edit.php]: Maybe a stupid question but why did you add this line? :-)

Comment (Bakual) [inline components/com_content/views/form/tmpl/edit.php]: Because the article form currently only renders a fixed set of fieldsets. But we need it to render all available fieldsets.
So I have changed it to it renders all except those which are already hardcoded rendered.

**Comment by coolcat-creations** [comment]
Please tag me if ready to test :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13353">issues.joomla.org/tracker/joomla-cms/13353</a>.</sub>

**Comment by Bakual** [comment]
@coolcat-creations It is ready for quite some time already :smile:

**Comment by ggppdk** [comment]
I will test too during weekend

**Comment by coolcat-creations** [comment]
@Bakual i have a conflict when trying to apply the patch in patchtester...

**Comment by coolcat-creations** [comment]
I have tested this item :white_check_mark: successfully on 2b5784ccd4371cbfb402e0104b6fbaaea66a953b<br /><br />@Bakual sorry i should have read my errormessage :-D Test successful!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13353">issues.joomla.org/tracker/joomla-cms/13353</a>.</sub>

**Comment by laoneo** [comment]
I will also have a look over the weekend, somehow I have the feeling something is missing.

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 2b5784ccd4371cbfb402e0104b6fbaaea66a953b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13353">issues.joomla.org/tracker/joomla-cms/13353</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC as there are 2 successfully Tests?

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13353">issues.joomla.org/tracker/joomla-cms/13353</a>.</sub>

**Comment by coolcat-creations** [comment]
@Bakual Before i could call single fields with  <?php $this->form->renderField('fieldname', 'params'); ?> by testing this patch i could call them with  <?php $this->form->renderField('fieldname', 'com_fields'); ?> but this does not work anymore. Do you have an idea why? (I´m testing in the contactform)
